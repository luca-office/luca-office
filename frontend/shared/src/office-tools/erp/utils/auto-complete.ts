import {ErpType} from "../../../enums"
import {ErpEntityByType, ErpKeyEntity} from "../../../models"
import {find} from "../../../utils"
import {UseErpForeignKeysHook} from "../hooks"

export const getAutoCompleteList = <T extends ErpType>(
  autoCompleteLists: TypeOf<UseErpForeignKeysHook, "foreignKeys">,
  foreignKeys: ErpKeyEntity<T>[],
  key: keyof ErpEntityByType<T>
): string[] =>
  find(foreignKey => foreignKey.key === key, foreignKeys)
    .map(foreignKey => autoCompleteLists[foreignKey.type] ?? [])
    .getOrElse([])
