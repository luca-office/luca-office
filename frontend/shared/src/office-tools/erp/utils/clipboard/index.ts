export * from "./clipboard"
export * from "./common"
export * from "./referencing-keys-clipboard"
