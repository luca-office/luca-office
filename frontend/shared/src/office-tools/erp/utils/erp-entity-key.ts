import {ErpType} from "../../../enums"
import {ErpEntity} from "../../../models"

const hasSalutation = (erpEntity: ErpEntity) => {
  switch (erpEntity.type) {
    case ErpType.Customer:
    case ErpType.Employee:
    case ErpType.Supplier:
      return true
    default:
      return false
  }
}

export const getErpEntityKeys = (erpEntity: ErpEntity): string[] => {
  const keys = Object.keys(erpEntity)

  // if entity has salutation it also contains firstName and lastName
  return hasSalutation(erpEntity)
    ? [
        "salutation",
        "firstName",
        "lastName",
        ...keys.filter(key => key !== "salutation" && key !== "firstName" && key !== "lastName")
      ]
    : keys
}
