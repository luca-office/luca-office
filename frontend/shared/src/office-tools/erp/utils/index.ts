export * from "./auto-complete"
export * from "./clipboard"
export * from "./common"
export * from "./erp-entity-key"
export * from "./erp-navigation-entry-label"
export * from "./input"
export * from "./label"
export * from "./number"
export * from "./referencing-keys"
export * from "./required"
export * from "./scenario-erp-entity"
export * from "./template"
