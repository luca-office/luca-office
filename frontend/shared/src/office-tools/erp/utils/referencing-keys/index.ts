export * from "./referencing-keys"
export * from "./referencing-keys-label"
