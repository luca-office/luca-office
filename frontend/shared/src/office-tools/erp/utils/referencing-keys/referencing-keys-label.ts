import {ErpType} from "../../../../enums"
import {LucaTFunction} from "../../../../translations"

// eslint-disable-next-line consistent-return
export const getAccordionLabel = (t: LucaTFunction, type: ErpType): string => {
  switch (type) {
    case ErpType.Component:
      return t("erp_accordion__component_label")
    case ErpType.Customer:
      return t("erp_accordion__customer_label")
    case ErpType.Employee:
      return t("erp_accordion__employee_label")
    case ErpType.Invoice:
      return t("erp_accordion__invoice_label")
    case ErpType.Order:
      return t("erp_accordion__order_label")
    case ErpType.OrderItem:
      return t("erp_accordion__order_item_label")
    case ErpType.Product:
      return t("erp_accordion__product_label")
    case ErpType.Supplier:
      return t("erp_accordion__supplier_label")
    case ErpType.ComponentProduct:
      return t("erp_accordion__component_for_product_label_simple")
  }
}

// eslint-disable-next-line consistent-return
export const getAdditionalAccordionLabel = (t: LucaTFunction, type: ErpType, count: number): string => {
  switch (type) {
    case ErpType.Component:
      return t("erp_accordion__component_additional_label", {count})
    case ErpType.Customer:
      return t("erp_accordion__customer_additional_label", {count})
    case ErpType.Employee:
      return t("erp_accordion__employee_additional_label", {count})
    case ErpType.Invoice:
      return t("erp_accordion__invoice_additional_label", {count})
    case ErpType.Order:
      return t("erp_accordion__order_additional_label", {count})
    case ErpType.OrderItem:
      return t("erp_accordion__order_item_additional_label", {count})
    case ErpType.Product:
      return t("erp_accordion__product_additional_label", {count})
    case ErpType.Supplier:
      return t("erp_accordion__supplier_additional_label", {count})
    case ErpType.ComponentProduct:
      return t("erp_accordion__component_for_product_additional_label", {count})
  }
}
