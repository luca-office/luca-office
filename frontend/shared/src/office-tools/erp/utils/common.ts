import {has, omit} from "lodash-es"
import {ErpNavigationEntryId, ErpType} from "../../../enums"
import {
  DeliveryStatus,
  EmploymentMode,
  FamilyStatus,
  PaymentStatus,
  Salutation
} from "../../../graphql/generated/globalTypes"
import {ErpEntity, ErpEntityByType, ErpKeyEntity, ErpStackEntity} from "../../../models"
import {LucaTFunction} from "../../../translations"
import {contains, removeTypename} from "../../../utils"

export const getForeignKeys = <T extends ErpType>(foreignKeys: ErpKeyEntity<T>[]): (keyof ErpEntityByType<T>)[] =>
  foreignKeys.map(foreignKey => foreignKey.key)

export const isForeignKey = <T extends ErpType>(
  foreignKeys: (keyof ErpEntityByType<T>)[],
  key: keyof ErpEntityByType<T>
): boolean => contains(key, foreignKeys)

// eslint-disable-next-line consistent-return
export const getSalutationLabel = (t: LucaTFunction, salutation: Salutation): string => {
  switch (salutation) {
    case Salutation.Mr:
      return t("erp__salutation_label_mr")
    case Salutation.Mrs:
      return t("erp__salutation_label_mrs")
    case Salutation.NonBinary:
      return t("erp__salutation_label_non_binary")
  }
}

// eslint-disable-next-line consistent-return
export const getEmploymentModeLabel = (t: LucaTFunction, mode: EmploymentMode): string => {
  switch (mode) {
    case EmploymentMode.Assistance:
      return t("erp__employment_mode_label_assistance")
    case EmploymentMode.FullTime:
      return t("erp__employment_mode_label_full_time")
    case EmploymentMode.PartTime:
      return t("erp__employment_mode_label_part_time")
    case EmploymentMode.Student:
      return t("erp__employment_mode_label_student")
  }
}

// eslint-disable-next-line consistent-return
export const getFamilyStatusLabel = (t: LucaTFunction, status: FamilyStatus): string => {
  switch (status) {
    case FamilyStatus.Divorced:
      return t("erp__family_status_label_divorced")
    case FamilyStatus.Married:
      return t("erp__family_status_label_married")
    case FamilyStatus.Single:
      return t("erp__family_status_label_single")
  }
}

// eslint-disable-next-line consistent-return
export const getPaymentStatusLabel = (t: LucaTFunction, status: PaymentStatus): string => {
  switch (status) {
    case PaymentStatus.Paid:
      return t("erp__payment_status_label_paid")
    case PaymentStatus.Unpaid:
      return t("erp__payment_status_label_unpaid")
  }
}

// eslint-disable-next-line consistent-return
export const getDeliveryStatusLabel = (t: LucaTFunction, status: DeliveryStatus): string => {
  switch (status) {
    case DeliveryStatus.Completed:
      return t("erp__delivery_status_label_completed")
    case DeliveryStatus.InProcess:
      return t("erp__delivery_status_label_in_process")
  }
}

export const columnValueListToString = (values: (string | number)[]) => values.join(", ")

export const toErpStackEntity = (
  data: ErpEntity,
  foreignKeys: ErpKeyEntity<typeof data.type>[],
  index: number
): ErpStackEntity => ({
  data: (has(data, "__typename") ? removeTypename(data as ErpEntity & {__typename: string}) : data) as ErpEntity,
  foreignKeys,
  index
})

export const cleanupStackEntity = (entity: ErpStackEntity) => ({
  ...entity,
  data: omit(entity.data, ["type", "binaryFile"])
})

export const navigationIdToType = (navigationId: ErpNavigationEntryId): ErpType | null => {
  switch (navigationId) {
    case ErpNavigationEntryId.Suppliers:
      return ErpType.Supplier
    case ErpNavigationEntryId.Customers:
      return ErpType.Customer
    case ErpNavigationEntryId.Products:
      return ErpType.Product
    case ErpNavigationEntryId.OrderItems:
      return ErpType.OrderItem
    case ErpNavigationEntryId.Offers:
      return ErpType.Order
    case ErpNavigationEntryId.Invoices:
      return ErpType.Invoice
    case ErpNavigationEntryId.Components:
      return ErpType.Component
    case ErpNavigationEntryId.ComponentsForProducts:
      return ErpType.ComponentProduct
    case ErpNavigationEntryId.StaffTable:
      return ErpType.Employee
    default:
      return null
  }
}

export const getAccordionLabelByKey = (t: LucaTFunction, key: string) => {
  switch (key) {
    case "id":
      return t("erp__table_label_id")
    case "componentId":
      return t("erp__table_label_component_id")
    case "productId":
      return t("erp__table_label_product_id")
    default:
      return key
  }
}
