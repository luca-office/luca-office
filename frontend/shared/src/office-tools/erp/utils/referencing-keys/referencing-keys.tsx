import {partial} from "lodash-es"
import * as React from "react"
import {ErpType} from "../../../../enums"
import {LucaTFunction} from "../../../../translations"
import {ErpDataSetAccordion} from "../../common/erp-data-set-overlay/erp-data-set-accordion/erp-data-set-accordion"
import {UseErpReferencingPrimaryKeysHook} from "../../hooks"
import {getReferencingKeysClipboardString} from "../clipboard"

export interface GetErpReferencingKeyAccordionsParams {
  readonly t: LucaTFunction
  readonly navigateToEntity: (erpType: ErpType, ids: number[]) => void
  readonly referencingPrimaryKeys: TypeOf<UseErpReferencingPrimaryKeysHook, "primaryKeys">
  readonly onCopyAll: (dataString: string) => void
  readonly onCopySingleToClipboard?: () => void
}

export const getErpReferencingKeyAccordions = ({
  t,
  navigateToEntity,
  referencingPrimaryKeys,
  onCopyAll,
  onCopySingleToClipboard
}: GetErpReferencingKeyAccordionsParams): React.ReactNode => {
  const types = Object.keys(referencingPrimaryKeys).map(key => key as ErpType)

  const navigateTo = (type: ErpType, ids: number[]) => navigateToEntity(type, ids)
  const handleCopyAll = (data: Record<string, number[]>) => onCopyAll(getReferencingKeysClipboardString(t, data))

  return types.length ? (
    <React.Fragment>
      {types.map((type, index) => {
        const data = referencingPrimaryKeys[type] ?? {}
        return (
          <ErpDataSetAccordion
            key={`erp_accordion_${type}_${index}`}
            type={type}
            data={data}
            onClick={partial(navigateTo, type)}
            onCopyToClipboard={onCopySingleToClipboard}
            onCopyAll={() => handleCopyAll(data)}
          />
        )
      })}
    </React.Fragment>
  ) : null
}
