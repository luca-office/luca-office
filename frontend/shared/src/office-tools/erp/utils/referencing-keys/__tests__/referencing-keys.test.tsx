import {shallow} from "enzyme"
import * as React from "react"
import {create} from "react-test-renderer"
import {erpInvoicesMockGraphQl} from "../../../../../../tests/__mocks__"
import {fakeTranslate} from "../../../../../../tests/utils/translate-mock"
import {ErpType} from "../../../../../enums"
import {ErpDataSetAccordion} from "../../../common/erp-data-set-overlay/erp-data-set-accordion/erp-data-set-accordion"
import {getErpReferencingKeyAccordions, GetErpReferencingKeyAccordionsParams} from "../referencing-keys"

const invoicesKeyList = erpInvoicesMockGraphQl.map(mock => mock.id)
const orderItemsKeyList = erpInvoicesMockGraphQl.map(mock => mock.id)

const defaultErpReferencingKeyAccordionsParams: GetErpReferencingKeyAccordionsParams = {
  t: fakeTranslate,
  navigateToEntity: jest.fn(),
  referencingPrimaryKeys: {[ErpType.Invoice]: {id: invoicesKeyList}, [ErpType.OrderItem]: {id: orderItemsKeyList}},
  onCopyAll: jest.fn()
}

describe("referencing-keys", () => {
  describe("getErpReferencingKeyAccordions", () => {
    it("renders correctly", () => {
      const component = create(
        getErpReferencingKeyAccordions(defaultErpReferencingKeyAccordionsParams) as React.ReactElement
      )
      const tree = component.toJSON()
      expect(tree).toMatchSnapshot()
    })
    it("has correct structure", () => {
      const component = shallow(<div>{getErpReferencingKeyAccordions(defaultErpReferencingKeyAccordionsParams)}</div>)
      expect(component.find(ErpDataSetAccordion)).toHaveLength(2)
    })
  })
})
