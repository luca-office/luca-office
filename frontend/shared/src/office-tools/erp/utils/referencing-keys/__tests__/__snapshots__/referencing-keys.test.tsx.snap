// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`referencing-keys getErpReferencingKeyAccordions renders correctly 1`] = `
[
  <div
    className="erp-accordion css-1bwtgc9-Card"
  >
    <div
      className="card-header css-1wbumr0-CardHeader"
    >
      <div
        className="css-17ett1w-AccordionCard"
        onClick={[Function]}
      >
        <div
          className="css-fsv1sy-AccordionCard"
        >
          <div
            className="css-v5hrte-AccordionCard"
          >
            erp_accordion__invoice_label (erp_dataset__foreign_key)
          </div>
          <div
            className="css-6s7adz-Icon"
            dangerouslySetInnerHTML={
              {
                "__html": {},
              }
            }
          />
        </div>
        <div
          className="css-w6fyje-AccordionCard"
        >
          erp_accordion__invoice_additional_label
        </div>
        <div
          className="css-l0jg45-Icon"
          dangerouslySetInnerHTML={
            {
              "__html": {},
            }
          }
        />
      </div>
    </div>
  </div>,
  <div
    className="erp-accordion css-1bwtgc9-Card"
  >
    <div
      className="card-header css-1wbumr0-CardHeader"
    >
      <div
        className="css-17ett1w-AccordionCard"
        onClick={[Function]}
      >
        <div
          className="css-fsv1sy-AccordionCard"
        >
          <div
            className="css-v5hrte-AccordionCard"
          >
            erp_accordion__order_item_label (erp_dataset__foreign_key)
          </div>
          <div
            className="css-6s7adz-Icon"
            dangerouslySetInnerHTML={
              {
                "__html": {},
              }
            }
          />
        </div>
        <div
          className="css-w6fyje-AccordionCard"
        >
          erp_accordion__order_item_additional_label
        </div>
        <div
          className="css-l0jg45-Icon"
          dangerouslySetInnerHTML={
            {
              "__html": {},
            }
          }
        />
      </div>
    </div>
  </div>,
]
`;
